### PR TITLE
CP-46631: Improved list of span attributes.

### PR DESCRIPTION
**Pull request by GabrielBuica**
Adds:
- `xs.xapi.task.id`;
- `xs.xapi.task.name`;
- `xs.xapi.task.uuid`;
- `xs.xapi.session.track.id`;
- `xs.xapi.task.origin`
to the list of span attributes created by `Context.make`.

This improves the debuggability and makes spans more easily identifiable. 

Also combines `complete_tracing` and `complete_tracing_with_exn` in one function by making error a optional argument.

### CODE REVIEW

File: ocaml/xapi/context.ml [hunk 1]
```diff
@@ -323,7 +325,16 @@ let make ?(http_other_config = []) ?(quiet = false) ?subtask_of ?session_id
             " by task " ^ make_dbg [] "" subtask_of
         )
   ) ;
-  let tracing = start_tracing_helper (parent_of_origin origin) task_name in
+  let span_attributes =
+    [
+      ("xs.xapi.task.id", Ref.really_pretty_and_small task_id)
+    ; ("xs.xapi.task.name", task_name)
+    ; ("xa.xapi.task.uuid", Uuidx.to_string task_uuid)
```
Comment (lindig) [inline ocaml/xapi/context.ml:332]: Typo: xa/xs

File: ocaml/xapi/context.ml [hunk 2]
```diff
@@ -234,17 +234,19 @@ let parent_of_origin (origin : origin) span_name =
   | _ ->
       None
 
-let start_tracing_helper parent_fn task_name =
+let start_tracing_helper ?(span_attributes = []) parent_fn task_name =
```
Comment (Vincent-lau) [inline ocaml/xapi/context.ml:237]: There are some other places that use `start_tracing_helper`, e.g. line 281. Do we need to update those as well?

File: ocaml/xapi/context.ml (hunk 1 above)
Comment (GabrielBuica) [reply ocaml/xapi/context.ml]: Thanks! Here it is https://github.com/xapi-project/xen-api/pull/5368/commits/5213d037487961bf2e26f720509b572e915f7480.

File: ocaml/xapi/context.ml (hunk 2 above)
Comment (GabrielBuica) [reply ocaml/xapi/context.ml:237]: This needs further discussion on what paths we want to instrument with this.

Comment (GabrielBuica) [reply ocaml/xapi/context.ml:255]: After discussing this with @mg12, I extended to list of new possible attributes and added them where appropriate to `start_tracing_helper`.

**Comment by GabrielBuica** [comment]
Passed bvt.

File: ocaml/xapi/context.ml [hunk 3]
```diff
@@ -234,17 +225,46 @@ let parent_of_origin (origin : origin) span_name =
...
+      (fun task_name -> [("xs.xapi.task.name", task_name)])
+      task_name
+  ; attribute_helper_fn
+      (fun task_id -> [("xs.xapi.task.id", Ref.really_pretty_and_small task_id)])
+      task_id
+  ; attribute_helper_fn
+      (fun task_uuid -> [("xs.xapi.task.uuid", Uuidx.to_string task_uuid)])
+      task_uuid
+  ; attribute_helper_fn
+      (fun session_id -> [("xs.xapi.session.id", Ref.string_of session_id)])
```
Comment (robhoes) [inline ocaml/xapi/context.ml:241]: You can't include session refs, as they are security (auth) tokens. You could translate them into track ids, as done for logging elsewhere in this module.

**Review by robhoes — APPROVED** [review]
Looks good. Please squash the fixup.

**Review by mg12 — APPROVED** [review]

**Pull request merged**